src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs

```diff
@@ -1432,5 +1434,50 @@ private void CheckDollarEachSegmentRestrictions(int index)
                 }
             }
         }
+
+        private static void CheckTypeCastSegmentRestrictions(IEdmModel model, ODataPathSegment previous, IEdmType targetEdmType)

```

Comment:
>CheckTypeCastSegmentRestrictions [](start = 28, length = 32)

Note; this doesn't support scenarios where the immediately previous segment was not a property or a navigation property;  for example, Customers/1/ns.PreferredCustomer.  Probably okay just to note this in the code.

Reply:
If the underlying type of the TypeDefinition is Edm.Primitive, then the DerivedTypeConstraint can be used to restrict the set of primitive types (i.e., to string and int).  If the DerivedTypeConstraint is defined on both a TypeDefinition and a Property where that DerivedTypeConstraint is used, then only those types that are in both lists are valid.

---
In reply to: [227070272](https://github.com/OData/odata.net/pull/1295#discussion_r227070272) [](ancestors = 227070272)

Reply:
@mikepizzo Thanks for the explanation. In this PR, i left some commented codes for type definition, because ODL doesn't support the Uri parse for the type definition, see issue at: https://github.com/OData/odata.net/issues/1326. So, i will un-comment type definition codes after i fix #1326. Is it ok?